migration/migration.go

```diff
@@ -1,25 +1,133 @@
 package migration
 
 import (
+	"log"
+
 	"github.com/almighty/almighty-core/app"
 	"github.com/almighty/almighty-core/models"
-	"github.com/almighty/almighty-core/remoteworkitem"
+	"github.com/almighty/almighty-core/transaction"
 	"github.com/jinzhu/gorm"
 	"golang.org/x/net/context"
 )
 
-// Perform executes the required migration of the database on startup
-func Perform(ctx context.Context, db *gorm.DB, witr models.WorkItemTypeRepository) error {
-	db.AutoMigrate(
-		&models.WorkItem{},
-		&models.WorkItemType{},
-		&remoteworkitem.Tracker{},
-		&remoteworkitem.TrackerQuery{},
-		&remoteworkitem.TrackerItem{})
-	if db.Error != nil {
-		return db.Error
+// getCurrentVersion returns the highest version from the version
+// table or -1 if that table does not exist.
+//
+// Returning -1 simplifies the logic of the migration process because
+// the next version is always the current version + 1 which results
+// in -1 + 1 = 0 which is exactly what we want as the first version.
+func getCurrentVersion(db *gorm.DB) (int, error) {
+	if !db.HasTable("version") {
+		return -1, nil
+	}
+
+	res := db.Raw("select max(version) from version")
+	if res.Error != nil {
+		return -1, res.Error
+	}
+
+	type VersionResult struct {
+		version int
+	}
+	var v VersionResult
+	res.Scan(&v)
+	return v.version, res.Error
+}
+
+// executeSQLFile loads the given fileName from the packaged SQL files and
+// executes it on the given database
+func executeSQLFile(db *gorm.DB, fileName string) error {
+	data, err := Asset(fileName)
+	if err != nil {
+		return err
+	}
+	return db.Exec(string(data)).Error
+}
+
+// Migrate executes the required migration of the database on startup
+func Migrate(db *gorm.DB) error {
+
+	type MigrationFunc func(tx *gorm.DB) error
+
+	migrations := [][]MigrationFunc{}
+
+	// Version 0
+	migrations = append(migrations, []MigrationFunc{
+		func(db *gorm.DB) error {
+			return executeSQLFile(db, "000-bootstrap.sql")
```

Comment:
executeSQLFile will probably be a very common migration strategy. Maybe create a 'FactoryFunction' that returns a MigrationFunc to avoid having to repeat the MigrationFunc anonymous definition each time. Something like:

```go
migrations = append(migrations, []MigrationFunc{ExecuteSQLFile("000-bootstrap.sql")})

func ExecuteSQLFile(file string) MigrationFunc {
   return func(db *gorm.DB) error {
      return executeSQLFile(db, file)
   }
}
```